plugins/qc-firehose/fu-qc-firehose-device.c

```diff
@@ -0,0 +1,1073 @@
+	if (!fu_qc_firehose_device_configure(self, "nand", error)) {
+		g_prefix_error(error, "failed to configure: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* each action */
+	xns = xb_silo_query(silo, "data/*", 0, error);
+	if (xns == NULL) {
+		g_prefix_error(error, "failed to get targets: ");
+		return FALSE;
+	}
+	if (!fu_qc_firehose_device_write_firmware_targets(self,
+							  firmware,
+							  xns,
+							  fu_progress_get_child(progress),
+							  error)) {
+		g_prefix_error(error, "failed to write targets: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* find the bootable partition */
+	if (self->supported_functions & FU_QC_FIREHOSE_FUNCTIONS_SETBOOTABLESTORAGEDRIVE) {
+		guint64 bootable = fu_qc_firehose_device_find_bootable(self, xns);
+		if (bootable != G_MAXUINT64) {
+			g_debug("setting partition %u bootable", (guint)bootable);
+			if (!fu_qc_firehose_device_set_bootable(self, (guint)bootable, error)) {
+				g_prefix_error(error, "failed to set bootable: ");
+				return FALSE;
+			}
+		}
+	}
+
+	/* reset, back to runtime */
+	if (self->supported_functions & FU_QC_FIREHOSE_FUNCTIONS_POWER) {
+		if (!fu_qc_firehose_device_reset(self, error)) {
+			g_prefix_error(error, "failed to reset: ");
+			return FALSE;
+		}
+		fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	}
+
+	/* success */
+	fu_device_remove_private_flag(FU_DEVICE(self), FU_QC_FIREHOSE_DEVICE_LOADED_SAHARA);
+	return TRUE;
+}
+
+static void
+fu_qc_firehose_device_parse_eps(FuQcFirehoseDevice *self, GPtrArray *endpoints)
+{
+	for (guint i = 0; i < endpoints->len; i++) {
+		FuUsbEndpoint *ep = g_ptr_array_index(endpoints, i);
+		if (fu_usb_endpoint_get_direction(ep) == FU_USB_DIRECTION_DEVICE_TO_HOST) {
+			self->ep_in = fu_usb_endpoint_get_address(ep);
+			self->maxpktsize_in = fu_usb_endpoint_get_maximum_packet_size(ep);
+		} else {
+			self->ep_out = fu_usb_endpoint_get_address(ep);
+			self->maxpktsize_out = fu_usb_endpoint_get_maximum_packet_size(ep);
+		}
+	}
+}
+
+static gboolean
+fu_qc_firehose_device_probe(FuDevice *device, GError **error)
+{
+	FuQcFirehoseDevice *self = FU_QC_FIREHOSE_DEVICE(device);
+	g_autoptr(GPtrArray) intfs = NULL;
+
+	/* most devices have a BCD version of 0.0 (i.e. unset), but we still want to show the
+	 * device in gnome-firmware -- allow overwriting if the descriptor has something better */
+	fu_device_set_version(device, "0.0");
+
+	/* FuUsbDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_qc_firehose_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* parse usb interfaces and find suitable endpoints */
+	intfs = fu_usb_device_get_interfaces(FU_USB_DEVICE(self), error);
+	if (intfs == NULL)
+		return FALSE;
+	for (guint i = 0; i < intfs->len; i++) {
+		FuUsbInterface *intf = g_ptr_array_index(intfs, i);
+		if (fu_usb_interface_get_class(intf) == 0xFF &&
+		    fu_usb_interface_get_subclass(intf) == 0xFF &&
+		    fu_usb_interface_get_protocol(intf) == 0xFF) {
```

Comment:
I also just found out that the protocol may not always be `0xFF` but sometimes is `0x11`.

Reply:
Fixed, thanks.